test/ctmrg/gradients.jl

```diff
@@ -2,12 +2,92 @@ using Test
 using Random
 using PEPSKit
 using TensorKit
-using PEPSKit
 using Zygote
 using OptimKit
+using ChainRulesCore
+using ChainRulesTestUtils
 using KrylovKit
+using FiniteDifferences
 
-# Square lattice Heisenberg Hamiltonian
+## Test utility
+# -------------
+function ChainRulesTestUtils.rand_tangent(rng::AbstractRNG, x::AbstractTensorMap)
+    return TensorMap(randn, scalartype(x), space(x))
+end
+function ChainRulesTestUtils.rand_tangent(rng::AbstractRNG, x::CTMRGEnv)
+    Ctans = x.corners
+    Etans = x.edges
+    for i in eachindex(x.corners)
+        Ctans[i] = rand_tangent(rng, x.corners[i])
+    end
+    for i in eachindex(x.edges)
+        Etans[i] = rand_tangent(rng, x.edges[i])
+    end
+    return CTMRGEnv(Ctans,Etans)
+end
+function ChainRulesTestUtils.test_approx(actual::AbstractTensorMap,
+                                         expected::AbstractTensorMap, msg=""; kwargs...)
+    for (c, b) in blocks(actual)
+        ChainRulesTestUtils.@test_msg msg isapprox(b, block(expected, c); kwargs...)
+    end
+end
+function ChainRulesTestUtils.test_approx(actual::InfinitePEPS,
+                                         expected::InfinitePEPS, msg=""; kwargs...)
+    for i in eachindex(size(actual,1))
+        for j in eachindex(size(actual,2))
+            ChainRulesTestUtils.@test_msg msg isapprox(actual[i,j], expected[i,j]; kwargs...)
+        end
+    end
+end
+function ChainRulesTestUtils.test_approx(actual::CTMRGEnv,
+    expected::CTMRGEnv, msg=""; kwargs...)
+    for i in eachindex(actual.corners)
+        ChainRulesTestUtils.@test_msg msg isapprox(actual.corners[i], expected.corners[i]; kwargs...)
+    end
+    for i in eachindex(actual.edges)
+        ChainRulesTestUtils.@test_msg msg isapprox(actual.edges[i], expected.edges[i]; kwargs...)
+    end
+end
+
+function FiniteDifferences.to_vec(t::T) where {T<:TensorKit.TrivialTensorMap}
+    vec, from_vec = to_vec(t.data)
+    return vec, x -> T(from_vec(x), codomain(t), domain(t))
+end
+function FiniteDifferences.to_vec(t::AbstractTensorMap)
+    vec = mapreduce(vcat, blocks(t)) do (c, b)
+        if scalartype(t) <: Real
+            return reshape(b, :) .* sqrt(dim(c))
+        else
+            v = reshape(b, :) .* sqrt(dim(c))
+            return vcat(real(v), imag(v))
+        end
+    end
+
+    function from_vec(x)
+        t′ = similar(t)
+        T = scalartype(t)
+        ctr = 0
+        for (c, b) in blocks(t′)
+            n = length(b)
+            if T <: Real
+                copyto!(b, reshape(x[(ctr + 1):(ctr + n)], size(b)) ./ sqrt(dim(c)))
+            else
+                v = x[(ctr + 1):(ctr + 2n)]
+                copyto!(b,
+                        complex.(x[(ctr + 1):(ctr + n)], x[(ctr + n + 1):(ctr + 2n)]) ./
+                        sqrt(dim(c)))
+            end
+            ctr += T <: Real ? n : 2n
+        end
+        return t′
+    end
+
+    return vec, from_vec
+end
```

Comment:
Have to recheck this with the current code from TensorKit, as this is copied from an old version. Ideally we move this to a package extension for ChainRulesTestUtils in TensorKit itself